076-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,389 @@
+3. Maybe triggering a rolling update of the ZooKeeper and Kafka nodes to trust a new Cluster CA certificate. This only happens if either:
+   1. The Cluster CA key got replaced during the current reconciliation, or
+   2. One or more Kubernetes Pods have the wrong strimzi.io/cluster-ca-key-generation annotation, meaning they do not trust the latest Cluster CA certificate.
+      This is only expected to occur if the operator previously updated the Cluster CA key and a rolling update was started during a previous reconciliation, but it was not completed.
+4. Maybe deleting an old CA certificate if it is not longer used
+   1. This only occurs if all pods in the cluster have the correct strimzi.io/cluster-ca-cert-generation annotation, meaning all the component certificates have been signed by the latest Cluster CA certificate.
+
+ZooKeeperReconciler and KafkaReconciler classes:
+1. Generate the certificates the Kafka and ZooKeeper nodes will present to clients (other ZK/Kafka nodes, the Cluster Operator, Kafka producer and consumers etc) and store them in the Kubernetes Secrets listed above, with the cluster-ca-cert-generation (and clients-ca-cert-generation for Kafka) annotation set to the current value passed in with the Cluster CA (and Clients CA).
+2. Create/update the StrimziPodSet used for the nodes in the cluster and add the following annotations:
+   1. cluster-ca-cert-generation set to the current values passed in with the Cluster CA.
+   2. cluster-ca-key-generation set to the current values passed in with the Cluster CA.
+   3. server-cert-hash set to the full SHA1-hash of the certificate that was generated.
+
+EntityOperatorReconciler, CruiseControlReconciler and KafkaExporterReconciler generate the certificates they will use to authenticate with Kafka and ZooKeeper and store them in the Kubernetes Secrets listed above, with the cluster-ca-cert-generation set to the current value passed in with the Cluster CA.
+
+### Example flows
+
+When a new cluster is first created the flow is:
+
+1. Either the user or the CaReconciler creates the Cluster CA certificate and key and stores them in the <CLUSTER_NAME>-cluster-ca-cert and <CLUSTER_NAME>-cluster-ca Kubernetes Secrets respectively.
+2. The CaReconciler generates the Cluster Operator EE certificate (signed by the newly generated Cluster CA certificate) and stores it in the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret.
+3. The component reconcilers (KafkaReconciler, ZooKeeperReconciler, EntityOperatorReconciler etc) generate their corresponding EE certificates (signed by the newly generated Cluster CA certificate), store them in Kubernetes Secrets and annotate the Kubernetes Secrets and Pods with the cert and key generations, and (for ZooKeeper and Kafka) the certificate hash.
+
+#### Fig 3: New cluster create flow
```

Comment:
It's because markdown doesn't provide a way to supply image captions, so this was my attempt at somehow captioning the image. I've updated it to instead put the images in a note box so hopefully that is easier to understand, rather than having the confusing headers.